var/spack/repos/builtin/packages/upcxx/package.py

```diff
@@ -40,12 +23,14 @@ class Upcxx(Package):
 
     homepage = "https://upcxx.lbl.gov"
 
+    version('2019.9.0', '7642877e05300e38f6fa0afbc6062788', default=True)
     version('2019.3.2', '844722cb0e8c0bc649017fce86469457')
 
     variant('cuda', default=False,
             description='Builds a CUDA-enabled version of UPC++')
 
-    variant('scheduler', values=('slurm', 'alps', 'user', 'none'),
+    variant('scheduler', values=('slurm', 'alps', 'none'),
```

Comment:
I was thinking about this, and I recall that @bonachea mentioned in a previous post that what really matters is that  `CROSS` is set.  Given that, why don't you just add a `cross` variant, give it a string value, and set it to the value  of the `CROSS` environment variable based on the detection you're doing?  This will satisfy a) the requirement for autodetection and b)  the requirement that everything be respected in the spec so that Spack knows exactly what was installed.

Reply:
I do think we can add a cross variant. What would you recommend for the accepted `values` and `default` ?

Reply:
More precisely, here is what I have tried:
```
--- a/var/spack/repos/builtin/packages/upcxx/package.py
+++ b/var/spack/repos/builtin/packages/upcxx/package.py
@@ -29,11 +29,11 @@ class Upcxx(Package):
     variant('cuda', default=False,
             description='Builds a CUDA-enabled version of UPC++')
 
-    variant('scheduler', values=('slurm', 'alps', 'none'),
+    variant('cross',default='cray-aries-%s'%detect_scheduler() 
+            if detect_scheduler() != 'none' else 'none',
+            description="Value for the CROSS environment variable")
 
-            default=detect_scheduler(),
-            description="Resource manager to use")
-    conflicts('scheduler=none', when='platform=cray',
+    conflicts('cross=none', when='platform=cray',
               msg='None is unacceptable on Cray.')
 
     depends_on('cuda', when='+cuda')
@@ -48,10 +48,11 @@ class Upcxx(Package):
     def setup_environment(self, spack_env, run_env):
         if 'platform=cray' in self.spec:
             spack_env.set('GASNET_CONFIGURE_ARGS', '--enable-mpi=probe')
-            if "scheduler=slurm" in self.spec:
-                spack_env.set('CROSS', 'cray-aries-slurm')
-            elif "scheduler=alps" in self.spec:
-                spack_env.set('CROSS', 'cray-aries-alps')
+
+        if not "cross=none" in self.spec:
+            #THIS IS FAILING NOW, how do I get the VALUE after "cross=XXXX" ?
+            spack_env.set('CROSS', self.spec['cross'])
+
         if '+cuda' in self.spec:
             spack_env.set('UPCXX_CUDA', '1')
             spack_env.set('UPCXX_CUDA_NVCC', self.spec['cuda'].prefix.bin.nvcc)
```

I have an issue when trying to set the CROSS value. Would you mind helping me on that?

Reply:
> Given that, why don't you just add a `cross` variant, give it a string value, and set it to the value of the `CROSS` environment variable based on the detection you're doing?

@tgamblin - if Mathias adds this `cross` variant, is it possible to default the string value based on `platform=cray` (itself another variant?) 

We need to maintain the property that the proposed variant defaults to `cross=none` (and leaves `$CROSS` unset) on non-Cray systems (even if they happen to have SLURM installed)

Reply:
Try `self.spec.variants['cross'].value`

Reply:
How can I do the following:
```
    variant('cross',default='cray-aries-%s'%detect_scheduler()
            if "platform=cray" in self.spec else 'none',
            description="Value for the CROSS environment variable")
```

The issue here is that I do not have access to `self`.

Reply:
@mjacquel: you can't do this in the general case (yet) as our current dependency solver isn't smart enough.   The `spec` isn't fully concrete when the default is set during concretization, so you can't rely on the platform being set.  It should eventually become possible to do something like this but it requires a full backtracking concretizer.

In this case, though, the platform is always known at package load time.  Try this:

```python
import spack.architecture

def cross-detect():
    if spack.architecture.platform() == 'cray':
        if which('srun'):
            return 'cray-aries-slurm'
        if which('alps'):
            return 'cray-aries-alps'
    return 'none'

class Upcxx(Package):
    variant('cross', values=str, default=cross_default(), 
            description="UPC++ cross-compile target (autodetect by default)")
```

Reply:
Note that this means your users can set CROSS explicitly like this:

```console
$ spack install upcxx  cross=some-new-platform
```